src/c_api/adapter.h

```diff
@@ -0,0 +1,478 @@
+      : col_ptr(col_ptr),
+        row_idx(row_idx),
+        values(values),
+        num_features(num_features) {}
+
+ private:
+  class Line {
+   public:
+    Line(size_t col_idx, size_t size, const unsigned* row_idx,
+         const float* values)
+        : col_idx(col_idx), size(size), row_idx(row_idx), values(values) {}
+
+    size_t Size() const { return size; }
+    COOTuple GetElement(size_t idx) const {
+      return COOTuple(row_idx[idx], col_idx, values[idx]);
+    }
+
+   private:
+    size_t col_idx;
+    size_t size;
+    const unsigned* row_idx;
+    const float* values;
+  };
+
+ public:
+  size_t Size() const { return num_features; }
+  const Line GetLine(size_t idx) const {
+    size_t begin_offset = col_ptr[idx];
+    size_t end_offset = col_ptr[idx + 1];
+    return Line(idx, end_offset - begin_offset, &row_idx[begin_offset],
+                &values[begin_offset]);
+  }
+
+ private:
+  const size_t* col_ptr;
+  const unsigned* row_idx;
+  const float* values;
+  size_t num_features;
+};
+
+class CSCAdapter : public detail::SingleBatchDataIter<CSCAdapterBatch> {
+ public:
+  CSCAdapter(const size_t* col_ptr, const unsigned* row_idx,
+             const float* values, size_t num_features, size_t num_rows)
+      : batch(col_ptr, row_idx, values, num_features),
+        num_rows(num_rows),
+        num_columns(num_features) {}
+  const CSCAdapterBatch& Value() const override { return batch; }
+
+  size_t NumRows() const { return num_rows; }
+  size_t NumColumns() const { return num_columns; }
+
+ private:
+  CSCAdapterBatch batch;
+  size_t num_rows;
+  size_t num_columns;
+};
+
+class DataTableAdapterBatch : public detail::NoMetaInfo {
+ public:
+  DataTableAdapterBatch(void** data, const char** feature_stypes,
+                        size_t num_rows, size_t num_features)
+      : data(data),
+        feature_stypes(feature_stypes),
+        num_features(num_features),
+        num_rows(num_rows) {}
+
+ private:
+  enum class DTType : uint8_t {
+    kFloat32 = 0,
+    kFloat64 = 1,
+    kBool8 = 2,
+    kInt32 = 3,
+    kInt8 = 4,
+    kInt16 = 5,
+    kInt64 = 6,
+    kUnknown = 7
+  };
+
+  DTType DTGetType(std::string type_string) const {
+    if (type_string == "float32") {
+      return DTType::kFloat32;
+    } else if (type_string == "float64") {
+      return DTType::kFloat64;
+    } else if (type_string == "bool8") {
+      return DTType::kBool8;
+    } else if (type_string == "int32") {
+      return DTType::kInt32;
+    } else if (type_string == "int8") {
+      return DTType::kInt8;
+    } else if (type_string == "int16") {
+      return DTType::kInt16;
+    } else if (type_string == "int64") {
+      return DTType::kInt64;
+    } else {
+      LOG(FATAL) << "Unknown data table type.";
+      return DTType::kUnknown;
+    }
+  }
+
+  class Line {
+    float DTGetValue(const void* column, DTType dt_type, size_t ridx) const {
+      float missing = std::numeric_limits<float>::quiet_NaN();
```

Comment:
So DT has fixed missing value?

Reply:
It has defined missing values for each type. This is separate from xgboost's missing value handling which is applied on top of this in the DMatrix constructor.